### PR TITLE
api: Issues/344 renewed upload handler

### LINKED ISSUE #344: api: upload handler
* Upload file to backend and automatically create a table that fits data (If that is possible)
* Upload file to backend into existing table.

/ui/databases/databaseName/schemas/schemaName/tables/tableName/upload POST

Entire payload as multipart/form-data
instead of json {
    Upload-file:  multipart/form-data
}

Temporary linking to wip branch
https://github.com/Embucket/icebucket/tree/issues/344_api_upload_handler

### PR DESCRIPTION
**Pull request by YaroslavLitvinov**
Closes #344 

* Added csv format parameters

### CODE REVIEW

**Comment by YaroslavLitvinov** [comment]
This pull request introduces substantial changes to the `ExecutionService`, `upload_file` API, error handling, and support for uploading data in multipart form-data format. Below is a detailed review of the updates:

---

### **Key Updates**

1. **`ExecutionService::upload_data_to_table` Enhancements**
   - **New Parameters:**
     - Added `format: Format` parameter to support different data formats.
     - Changed return type to `ExecutionResult<usize>` to return the number of rows loaded.
   - **Unique Table Names:**
     - Temporary table names now include a UUID to handle simultaneous uploads.
   - **Error Handling:**
     - Replaced direct error mappings with `Snafu` error contexts for better error reporting.
   - **Row Tracking:**
     - Tracks the number of rows loaded during the upload operation.

2. **Error Handling Improvements**
   - Introduced new error variants in `TableError` and `TablesAPIError` for better granularity:
     - `MalformedMultipart`
     - `MalformedMultipartFileData`
     - `FileField`
     - Transparent forwarding of execution and metastore errors.
   - `status_code` mapping for `TablesAPIError` has been updated for more precise HTTP status codes.

3. **New `upload_file` Endpoint**
   - Implements an API to upload data to tables via multipart form-data:
     - Parses `UploadParameters` for CSV characteristics like header, delimiter, quote, etc.
     - Processes uploaded files using `datafusion::arrow::csv::reader::Format`.
     - Tracks the number of rows uploaded and reports it in the response.
   - Includes detailed responses for various error scenarios (e.g., malformed requests, missing fields).

4. **Testing Updates**
   - Updated tests in `service.rs` to reflect the new method signature and return type.
   - Added assertions for the number of rows loaded (`assert_eq!(rows_loaded, 3)`).

5. **Router Changes**
   - Added a new route for uploading data: `/databases/{databaseName}/schemas/{schemaName}/tables/{tableName}/rows`.

6. **Models Update**
   - Added models for handling upload payloads and responses:
     - `TableUploadPayload`: Defines the request schema for file uploads.
     - `TableUploadResponse`: Captures the number of rows loaded and duration.
   - `UploadParameters` now maps CSV characteristics into `datafusion::arrow::csv::reader::Format`.

---

### **Strengths**

1. **Improved API Usability:**
   - The new multipart form-data upload endpoint makes the API more user-friendly and integrates well with modern frontends.

2. **Enhanced Error Handling:**
   - The use of `Snafu` for error contexts simplifies error propagation and makes debugging easier.

3. **Concurrency Support:**
   - Generating unique names for temporary tables solves potential conflicts during simultaneous uploads.

4. **Better Testing:**
   - Tests now verify the number of rows uploaded, ensuring correctness.

5. **Code Modularity:**
   - The introduction of `UploadParameters` and its conversion to `Format` improves separation of concerns.

---

### **Concerns**

1. **Potential Performance Bottleneck:**
   - Using a UUID for each upload adds overhead to table creation and cleanup. Consider whether this is the best approach for high-throughput systems.

2. **Limited Validation for Parameters:**
   - The `UploadParameters` struct lacks validation for some fields (e.g., invalid delimiters, escape characters). This could lead to runtime errors.

3. **Error Propagation in `upload_file`:**
   - If multiple fields in the multipart form-data request fail, only the first error is reported. Aggregating errors might improve debugging.

4. **Backward Compatibility:**
   - The changes to `upload_data_to_table`'s method signature might break existing callers. Ensure that all dependent code is updated accordingly.

5. **Testing Scope:**
   - While the tests check for rows loaded, additional scenarios (e.g., malformed requests, invalid parameters) should be tested.

---

### **Recommendations**

1. **Optimize UUID Usage:**
   - Consider alternative approaches to handling simultaneous uploads (e.g., a pool of temporary table names or hashed identifiers).

2. **Validate Input Parameters:**
   - Add validation to `UploadParameters` to ensure all fields are valid before processing.

3. **Extend Test Coverage:**
   - Add tests for edge cases like missing headers, invalid delimiters, or malformed multipart requests.

4. **Improve Documentation:**
   - Provide detailed examples for the new endpoint, including how to structure multipart form-data requests.

5. **Handle Aggregated Errors:**
   - Enhance error reporting in `upload_file` to include all issues encountered during multipart parsing.

---

### **Conclusion**

This pull request significantly enhances the upload functionality and error handling capabilities of the service. While there are some areas for improvement (e.g., input validation, performance optimization), the changes are overall well-structured and align with modern API practices. Addressing the concerns and recommendations above will further strengthen the implementation.

File: crates/runtime/src/http/ui/tables/handlers.rs [hunk 1]
```diff
@@ -266,6 +277,95 @@ pub async fn get_table_preview_data(
...
+)]
+#[tracing::instrument(level = "debug", skip(state, multipart), err, ret(level = tracing::Level::TRACE))]
+pub async fn upload_file(
+    DFSessionId(session_id): DFSessionId,
+    Query(parameters): Query<UploadParameters>,
+    State(state): State<AppState>,
+    Path((database_name, schema_name, table_name)): Path<(String, String, String)>,
+    mut multipart: Multipart,
+) -> TablesResult<Json<TableUploadResponse>> {
+    let mut uploaded = false;
```
Comment (rampage644) [inline crates/runtime/src/http/ui/tables/handlers.rs:318]: Is this needed? Had we have any error we would exit by the time `uploaded` set to true

**Review by rampage644 — APPROVED** [review]
minor comment

File: crates/runtime/src/http/ui/tables/handlers.rs (hunk 1 above)
Comment (YaroslavLitvinov) [reply crates/runtime/src/http/ui/tables/handlers.rs:318]: It should fail If form data has no fields
